server/interpreter.js

```diff
@@ -325,11 +305,15 @@ Interpreter.prototype.step = function() {
   try {
     var nextState = stepFuncs_[node['type']].call(this, stack, state, node);
   } catch (e) {
-    // Eat any step errors.  They have been thrown on the stack.
-    if (e !== Interpreter.STEP_ERROR) {
+    if (e instanceof Error) {
       // Uh oh.  This is a real error in the interpreter.  Rethrow.
       throw e;
+    } else if (typeof e !== 'boolean' && typeof e !== 'number' &&
+        typeof e !== 'string' && e !== undefined && e !== null &&
+        !(e instanceof this.Object)) {
```

Comment:
Performance: Reorder these from most- to least-likely.

Reply:
Hard to say which is more likely when none of them should ever occur unless there is a defect elsewhere in the interpreter.

Reply:
I thought about this further, and realised my reply was completely wrong.

The most common case is that the user has thrown a `this.Error`, which is a `this.Object`, so short circuit that case immediately.  Then short circuit if `e === null`, since that's also a valid exception value.  Then invert the other four primitive tests and instead check for `typeof e === 'object' || typeof e === 'function'`, which (having dealt with real `Errors` and all pseudo `Object`s above), shouldn't happen.